### ISSUE TITLE
Support parsing loaded elf?

### ISSUE THREAD

**Issue by RisinT96** [gimli-rs/object#557]
Hello, 

I have a use case where I want to parse an elf that's already loaded into my processes memory, as the elf is loaded it's missing sections - which the current elf parser is heavily reliant upon.

Some use could still be had from parsing the dynamic segment (PT_DYNAMIC) to find dynamic symbols, relocations, etc...

I have something similar implemented, using object's type system, but I would like to add that as a built-in feature of this project.

Is this something that could interest you?

Thank you!

**Comment by philipc** [comment]
I'm interested in it. This sounds similar to #548. I had thought that this sort of thing wasn't possible for ELF, so it'd be good to see what you have done.

**Comment by RisinT96** [comment]
So `section_by_index` and `symbol_by_index` should probably be changed to return `Result<Option<_>>`.

**Comment by philipc** [comment]
Hmm, not sure I like changing those APIs either. How about try to do it without API changes and we can see what it looks like.

**Comment by RisinT96** [comment]
Sure

I'll have some time to work on this on the weekend.

**Comment by RisinT96** [comment]
I have an update,

When parsing the dynamic table, reading an entry that contains an address (for example `DT_JMPREL` which should point to `.rela.plt`), I'm getting different results on different machines:
* On Android devices (both 32 and 64 bit) I'm getting an offset into `data`
* On Linux machines (ubuntu 20.04 amd64, wsl2 ubunru 22.04 amd64) I'm getting the actual address of `.rela.plt`

I believe it has something to do with android having a custom dynamic linker than regular Linux.

I'm wondering what would be the correct way to handle this.

Perhaps, since this is only useful when a process is analyzing itself/other processes on the same system, it would make sense to use some kind of cfg depending on the compiled target, and add a workaround for all Android targets.

Or some kind of heuristic on the returned value, i.e. if it exceeds the size of data or something.

I'm not sure if there's a sure way to know :/

**Comment by philipc** [comment]
This may be getting into the kind of reason why I thought this wasn't possible for ELF. If you're relying on undocumented internals of dynamic linkers, then that's an area that I'm not sure I want to support in the high level API of this crate. It may be better to adapt the low level API to suit your needs instead.

Can you provide code to show how you are using this?

**Comment by RisinT96** [comment]
You can see a rough and dirty draft [here](https://github.com/gimli-rs/object/compare/master...RisinT96:object:feature/support-parsing-loaded-elf)

It successfully finds all the relevant dynamic sections.

It's useful for me when trying to parse the dynamic relocations and PLT of another .so loaded into my process, I can find its base address using dladdr, and through parsing its ELFs header (go over all program headers, find the segment with largest `vadrr + memsz`) I can find its size.

This allows me to get a slice that encapsulates the entire elf.

**Comment by philipc** [comment]
> You can find a very rough draft here

Thanks.

>  I can find its base address using dladdr, and by parsing its ELFs header (go over all program headers, find the segment with largest vadrr + memsz) I can find its size.

Are you able to share the code for that too, and ideally an example of the sort of thing you want to do with the resulting parsed file. We'll need to have a test of some kind in this crate.

Getting back to your question in https://github.com/gimli-rs/object/issues/557#issuecomment-1621132435, my feeling is that the ideal solution would result in the compiled executable including code that only supports the compiled target, and thus can work reliably instead of using heuristics.

If I were to write this code myself, instead of using the higher level API (things like `File`, `ElfFile`, and `ElfSymbolTable`), I would use the lower level API (things like `elf::ProgramHeader` and `read::elf::SymbolTable`), and use cfgs in my own code to control behaviour that depends on the compiled target. That is, I would write code that looks similar to the `ElfFile::parse_loaded` that you have written, but that code would live in my own crate. This may involve improving the lower level API in `object` to make it easier to do that.

**Comment by RisinT96** [comment]
> Are you able to share the code for that too, and ideally an example of the sort of thing you want to do with the resulting parsed file. We'll need to have a test of some kind in this crate.
I'll see what I can share

> Getting back to your question in https://github.com/gimli-rs/object/issues/557#issuecomment-1621132435, my feeling is that the ideal solution would result in the compiled executable including code that only supports the compiled target, and thus can work reliably instead of using heuristics.
Yeah that's what I thought as well, use cfgs that use offsets when compiling an Android target, and cfgs that use absolute addresses on Linux targets, and disable this entirely on untested targets

> If I were to write this code myself, instead of using the higher level API (things like `File`, `ElfFile`, and `ElfSymbolTable`), I would use the lower level API (things like `elf::ProgramHeader` and `read::elf::SymbolTable`), and use cfgs in my own code to control behaviour that depends on the compiled target. That is, I would write code that looks similar to the `ElfFile::parse_loaded` that you have written, but that code would live in my own crate. This may involve improving the lower level API in `object` to make it easier to do that.
That's basically what I already have implemented, unfortunately I can't share that code as is.

I was hoping to implement some of that functionality directly into this library for easier access for everyone.

**Comment by philipc** [comment]
I think adding code in the lower level API that is similar to `ElfFile::parse_loaded` could be useful for others, but I'm not convinced that `ElfFile` itself should support this.

For your use case, are you doing things that are specific to ELF? If we added `ElfFile::parse_loaded` and `PeFile::parse_loaded`, would your application be able to work on Windows by changing the `ElfFile::parse_loaded` call to a `PeFile::parse_loaded` call? If your code can't be written to work on both, then there is no reason to use an abstraction layer.

**Comment by RisinT96** [comment]
I'm not familiar enough with PE, but from a short search it seems like PEs IAT is very similar to ELFs PLT, and can be modified to the same effect.

I imagine the dynamic relocations abstraction gives access to the same information.

In conclusion, yes, I think there's value to be had in the abstraction.

**Comment by bjorn3** [comment]
dl_iterate_phdr returns a program header for every loaded DSO. According to the man page you have to calculate absolute addresses of segments using `info->dlpi_addr + info->dlpi_phdr[x].p_vaddr`. Does this work on both glibc and bionic (android) and does it return absolute addresses in both cases?

**Comment by RisinT96** [comment]
> dl_iterate_phdr returns a program header for every loaded DSO. According to the man page you have to calculate absolute addresses of segments using `info->dlpi_addr + info->dlpi_phdr[x].p_vaddr`. Does this work on both glibc and bionic (android) and does it return absolute addresses in both cases?

I'm getting absolute addresses in both cases

**Comment by RisinT96** [comment]
So I decided to start slow, added some lower level logic for parsing the dynamic segment.

This should work both if the segment was found using `dl_iterate_phdr`, or by manually parsing a loaded elf.

let me know what you guys think.

**Comment by bjorn3** [comment]
Parsing the dynamic segment is as easy as calling `ReadRef::read_slice::<Dyn64>()` on a slice with the segment data (or `Dyn32` for an ELF32 file). This gives you a `&[Dyn64]` with the individual entries. See the `dynamic` method on several of the types in the elf read module. You can then use the `Dyn` trait for convenience functions on the dynamic segment entries.

**Comment by RisinT96** [comment]
> Edit: Just saw your PR. Looks like you were thinking about a high level parsing interface, while I was thinking about a low level one. A high level one is of course nicer for most use cases.

Yeah, by low level I meant it was lower level than parsing the entire elf from scratch.

It's probably closer to high level than low level though 😅
